src/diskusage.cpp

```diff
@@ -0,0 +1,176 @@
+ * Contact: Thomas Perl <thomas.perl@jolla.com>
+ *
+ * You may use this file under the terms of the BSD license as follows:
+ *
+ * "Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Nemo Mobile nor the names of its contributors
+ *     may be used to endorse or promote products derived from this
+ *     software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
+ */
+
+#include "diskusage.h"
+#include "diskusage_p.h"
+
+#include <QThread>
+#include <QDir>
+#include <QProcess>
+
+DiskUsageWorker::DiskUsageWorker(QObject *parent)
+    : QObject(parent)
+{
+}
+
+DiskUsageWorker::~DiskUsageWorker()
+{
+}
+
+static quint64 calculateSize(const QString &directory)
+{
+    if (!QDir(directory).exists()) {
+        return 0L;
+    }
+
+    // TODO: Recursively sum up disk usage manually while avoiding symlinks
+    // instead of exec'ing du for every app
+
+    QProcess du;
+    du.start("du", QStringList() << "-sb" << directory, QIODevice::ReadOnly);
+    du.waitForFinished();
+    QStringList size_directory = QString::fromUtf8(du.readAll()).split('\t');
+
+    if (size_directory.size() > 1) {
+        return size_directory[0].toULongLong();
+    }
+
+    return 0L;
+}
+
+void DiskUsageWorker::submit(QStringList paths)
+{
+    foreach (const QString &path, paths) {
+        // TODO: Detect and handle errors (emit error(path, "..."))
+        emit calculated(path, calculateSize(path));
+    }
+
+    emit finished();
+}
+
+
+class DiskUsagePrivate
+{
+    Q_DISABLE_COPY(DiskUsagePrivate)
+    Q_DECLARE_PUBLIC(DiskUsage)
+
+    DiskUsage * const q_ptr;
+
+public:
+    DiskUsagePrivate(DiskUsage *usage);
+    ~DiskUsagePrivate();
+
+private:
+    QThread m_thread;
```

Comment:
Whether to or not to include specific subdirectories' size to parent sounds like a UI detail that might be wanted either way. If wanted, I'd expect it to be simpler on QML side to subtract size reported by other DiskUsage elements than to parse this list into proxy model. Alternative of supporting a list of directories to avoid should be simple too.